usecases/traverser/explorer_hybrid.go

```diff
@@ -0,0 +1,320 @@
+	//
+	// 1. If the user hase provided nearText parameters, use them in a nearText search
+	// 2. If the user has provided nearVector parameters, use them in a nearVector search
+	// 3. (Default) Do a vector search with the default parameters (the old hybrid search)
+	if (params.HybridSearch.Alpha) > 0 {
+		if params.HybridSearch.NearTextParams != nil {
+			res, name, err := nearTextSubSearch(ctx, e, params)
+			if err != nil {
+				e.logger.WithField("action", "hybrid").WithError(err).Error("nearTextSubSearch failed")
+				return nil, err
+			} else {
+				weights = append(weights, params.HybridSearch.Alpha)
+				results = append(results, res)
+				names = append(names, name)
+			}
+		} else if params.HybridSearch.NearVectorParams != nil {
+			res, name, err := denseSearch(ctx, params.HybridSearch.NearVectorParams, e, params, "nearVector")
+			if err != nil {
+				e.logger.WithField("action", "hybrid").WithError(err).Error("denseSearch failed")
+				return nil, err
+			}
+			weights = append(weights, params.HybridSearch.Alpha)
+			results = append(results, res)
+			names = append(names, name)
+		} else {
+			sch := e.schemaGetter.GetSchemaSkipAuth()
+			class := sch.FindClassByName(schema.ClassName(params.ClassName))
+			if class == nil {
+				return nil, fmt.Errorf("class %q not found", params.ClassName)
+			}
+
+			vectoriser := class.Vectorizer
+			if vectoriser != "none" {
+				if len(params.HybridSearch.Vector) == 0 {
+					var err error
+					params.SearchVector, err = e.modulesProvider.VectorFromInput(ctx, params.ClassName, params.HybridSearch.Query, targetVector)
+					if err != nil {
+						return nil, err
+					}
+				} else {
+					params.SearchVector = params.HybridSearch.Vector
+				}
+
+				// Build a new vearvec search
+				nearVecParams := &searchparams.NearVector{
+					Vector:        params.SearchVector,
+					TargetVectors: params.HybridSearch.TargetVectors,
+				}
+
+				res, name, err := denseSearch(ctx, nearVecParams, e, params, "hybridVector")
+				if err != nil {
+					e.logger.WithField("action", "hybrid").WithError(err).Error("denseSearch failed")
+					return nil, err
+				} else {
+					weights = append(weights, params.HybridSearch.Alpha)
+					results = append(results, res)
+					names = append(names, name)
+				}
+			}
+		}
+	}
+
+	// If the user has given any weight to the keyword search, do a keyword search
+	if 1-params.HybridSearch.Alpha > 0 {
+		sparseResults, name, err := sparseSearch(ctx, e, params)
+		if err != nil {
+			e.logger.WithField("action", "hybrid").WithError(err).Error("sparseSearch failed")
+			return nil, err
+		} else {
+			weights = append(weights, 1-params.HybridSearch.Alpha)
+			results = append(results, sparseResults)
+			names = append(names, name)
+		}
+	}
+
+	// The postProcess function is used to limit the number of results and to resolve references
+	// in the results.  It is called after all the subsearches have been completed, and before autocut
```

Comment:
more out of interest - why would we resolve references here?

Reply:
The next internal stage uses search results, and aggregate has a different post-process function.  It could probably be reworked a bit